lsp/nls/src/requests/completion.rs

```diff
@@ -716,15 +716,24 @@ pub fn handle_completion(
     id: RequestId,
     server: &mut Server,
 ) -> Result<(), ResponseError> {
-    // -1 because at the current cursor position, there is no linearization item there,
-    // so we try to get the item just before the cursor.
-    let mut pos = server.cache.position(&params.text_document_position)?;
-    pos.index.0 -= 1;
+    // The way indexing works here is that if the input file is
+    //
+    // foo‸
+    //
+    // then the cursor (denoted by ‸) is at index 3 and the span of foo is [0,
+    // 3), which does not contain the cursor. For most purposes we're interested
+    // in querying information about foo, so to do that we use the position just
+    // *before* the cursor.
```

Comment:
does this mean that if the cursor is over the first letter of a variable name, we won't get completion for that variable (unless it's the first character of the line)?

Reply:
I think it should be fine, basically because we're using the surrounding context to provide completion; we don't actually care about the identifier under the cursor.

To be specific, In the context of a record path you should get completion, because if the input is `foo.bar` and the cursor is "over" the `b` and so you'll get the list of completions implied by `foo.`. (This is the usual behavior for LSP; the editor is in charge of doing any filtering based on prefixes etc.)

In the context of a "bare" variable name, like the `foo` in `{ foo = 1 } | { foobar | Number }` it should also be fine, because we'll see that we're inside a record and use that record's fields and annotations for completion.